common/kvstore/src/main/java/org/apache/spark/util/kvstore/LevelDB.java

```diff
@@ -247,17 +266,28 @@ public void close() throws IOException {
     }
   }
 
+  public boolean isClosed() {
+    return this._db.get() == null;
+  }
+
   /**
    * Closes the given iterator if the DB is still open. Trying to close a JNI LevelDB handle
    * with a closed DB can cause JVM crashes, so this ensures that situation does not happen.
    */
-  void closeIterator(LevelDBIterator<?> it) throws IOException {
+  public void closeIterator(DBIterator it) throws IOException {
     synchronized (this._db) {
       DB _db = this._db.get();
       if (_db != null) {
         it.close();
       }
     }
+    iteratorTracker.remove(it);
+  }
+
+  public DBIterator createIterator() {
```

Comment:
I don't quite get this change - this was already managed correctly inside the LevelDBIterator. I don't think that was the issue. The previous behavior seemed fine for closeIterator

Reply:
`LevelDB` needs to close all active `DBIterator` instances in it's `close` function. So `LevelDB` needs get notify when `DBIterator` is created or closed and add /remove them from `iteratorTracker`. yes, here should be no other changes except this behavior. i also renamed `createIterator` to avoid confusion.

Reply:
But that is all managed already by LevelDBIterator. If they are being closed, the internal DBIterator is being closed.

Reply:
in my before commits, `LevelDBIterator` are tracked in `LevelDB` by `iteratorTracker`  to make sure they are correctly closed. I found GC would not collect `LevelDBIterator` objects if caller doesn't explicit close it (remove it from `iteratorTracker` ). So here is the change, let `LevelDB` tracks `DBIterator`. Then GC sill works as before. when `LevelDBIterator.finalize` is executed, `DBIterator` can be closed and removed from `iteratorTracker` .
As `DBIterator` would be closed in `LevelDB.close`, i think it's better to keep one class to manage it. Now `LevelDBIterator.finalize` and `LevelDBIterator.close` both finally calls `db.closeIterator`.
Thanks.

Reply:
But if you're tracking LevelDBIterator and closing them, you are already closing DBIterators inside. That was not the issue. This just makes it more complex.

Reply:
yes, you are right, this is a little complex. the issue here is if tracking `LevelDBIterator`, and some callers don't explicitly close it, `LevelDBIterator.finalize` can't be executed as `LevelDBIterator` are referenced in `LevelDB`. There may be the case elements in `iteratorTracker` may always increase but not get released till DB is closed. i think this would introduce memory leak in this case.
If all callers explicitly close `LevelDBIterator`, then we can track `LevelDBIterator` here.

Reply:
I see your point now about finalize. Well, I'd leave it in anyway.
But, tracking DBIterator doesn't improve things; you still don't end up closing them until the DB is closed. Right?

Reply:
I tried soft reference and weak reference,  these two would not help with leaking issue: they would be marked as null, when finalize ready. so here maybe the case GC and DB.close are executed same time. If the execution sequence is 1) weak reference is marked null, 2) `DB.close` then 3) `LevelDBIterator.finalize`. JNI resource still is leaked (`ReferenceQueue` also can't help as there is no guarantee for timing). UT i added can reproduce this.
`DBIterator` is internal property, it would not block `LevelDBIterator` be collected by GC. Then `LevelDBIterator.finalize` would close `DBIterator` and remove it from `iteratorTracker`.

Reply:
Yes, so why track DBIterator separately?
If an iterator leaks, previously, it might be closed by finalize(). If you hold a strong ref to all iterators, they can't be GCed. A soft reference would allow it. If the reference is null, that's fine, it means it was GCed and thus finalized and closed. Weak refs won't work here, yes.

Reply:
> Yes, so why track DBIterator separately?
> If an iterator leaks, previously, it might be closed by finalize(). 

If an iterator leaks (i mean after db is closed), it would not be closed any more. Here is the cause: If level db is closed, access to related JNI handle ( 
calling `DBIterator.close()` here ) would trigger JVM crashes. This is also the reason that we need check if db is closed in `LevelDBcloseIterator`.
![4](https://user-images.githubusercontent.com/10524738/84468475-7f1cbc80-acb1-11ea-8dde-5b3775463fbb.jpg)
This means if db is closed, then `LevelDBIterator.finalize()` will not close `DBIterator`. `DBIterator` and related JNI resource will be leaked till restarting.
 Yes, `DBIterator` are held, if DB is opend, `LevelDBIterator.finalize()` or `LevelDBIterator.close` would close and release them. otherwise they would be closed in `LevelDB.close`. This makes sure they can be closed and released in time.

> If you hold a strong ref to all iterators, they can't be GCed. A soft reference would allow it. If the reference is null, that's fine, it means it was GCed and thus finalized and closed. Weak refs won't work here, yes.

About soft reference, I think it has same issue with weak reference: it's value ( `get()` ) is marked as null before `finalize` get executed, this behavior is same with weak reference. maybe not too often, but it's still has chance encounter the race condition i mentioned above.

Reply:
I tested the case thoroughly by history server, soft reference is totally ok here. and it can also simple the lgoic. i updated PR to use it instead of `DBIterator`. @srowen thanks for you comments, Could you please take a look again?